File: C++/94.binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    vector<int> result;

    void inorder_traverse(TreeNode* root){
        if (root == NULL) {
            return;
        }

        inorder_traverse(root->left);
        result.push_back(root->val);
        inorder_traverse(root->right);

    }

    vector<int> inorderTraversal(TreeNode* root) {

        inorder_traverse(root);
        return result;

    }
};

File: 94.binary-tree-inorder-traversal.py
# Definition for a binary tree node.
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ## left -> root -> right
        result = list()

        def traverse(root):
            if root is None:
                return

            ## first recursively go to the left most child
            traverse(root.left)
            ## append the value
            result.append(root.val)
            ## now traverse the right most child
            traverse(root.right)

        traverse(root)

        return result